### ISSUE TITLE
feat(analysis): proper static EDD usage analysis (EL/entity-stack aware, with enumeration-bounded dynamic strings)

### ISSUE THREAD

**Issue by PaulSnow** [DTRules/DTRules#776]
## Problem

`pkg/dtrules/analysis/edd_unused.go` is a regex pass over DSL text:

```go
var identifierPattern = regexp.MustCompile(`\b([a-z_][a-z0-9_]*)\.([a-z_][a-z0-9_]*)\b`)
```

It only matches dotted `entity.attribute` references. Bare field names that resolve against the runtime entity stack (`for all accounts where identity_url == current_delegate_url` → `identity_url` resolves to `account.identity_url`) are invisible to the analyzer. Consequence: every field accessed via the bare-name path is flagged as "unused EDD field," producing systemic false positives.

A downstream consumer (staking) reports **97 EDD-usage warnings** on their project, the vast majority of which are this class of false positive.

## Scope

Replace the regex pass with proper static analysis over the EL AST. Walk every `*_dt.xml`, parse each DSL fragment with `pkg/dtrules/compiler/el`, and track the entity stack as the walk descends. Resolve bare identifiers against the topmost entity at each reference site.

**Entity-stack pushes the analyzer needs to track:**

- `for all <field>` context — pushes the iterated entity type for the duration of the table.
- `for all <field> where <pred>` / `for first of <field> where <pred>` — same, plus the `where` clause is in scope of the iterated entity.
- `using <entity> { ... }` blocks — pushes a named entity.
- `create <type> as <local>` (#764-era Gap A) — binds a local; field accesses against the local are dotted, but `set local.field = expr` is in scope of the local on the entity stack while the create-block runs.
- `createentity` / `entitypush` postfix idioms in the legacy hand-coded paths — should produce a conservative "unknown entity pushed" marker (see below).
- `perform <Table>` — descends into the called table's context; the analyzer needs cross-table reference tracking.

**Out of scope for the first cut:**

- Run-time deduction of which entity is on the stack inside an arbitrary postfix block. The strict hand-coded-postfix gate (shipped in v1.11.0) means new code can't introduce hand-edited postfix anyway.
- Closure-over-locals analysis (allocate / local@ / local!). Locals are typed at point of definition by the EL compiler; the analyzer can read those type annotations rather than reason about them.

## Dynamic-string corner case (enumeration bounds)

`perform table named (<string-expression>)` (shipped v1.12.0) lets a rule pick a table by computed name at runtime. Static analysis sees the string expression but can't enumerate its possible values without help. Same problem if we ever ship `entitypush <string>` — the analyzer wouldn't know which entity types could land on the stack.

Proposed solution: let authors **bound the string** with an enumeration the analyzer can read.

### Syntax sketch — enumeration-bounded perform

EDD declaration:

```xml
<entity name=\"job\">
  <field name=\"state\" type=\"string\">
    <enumeration kind=\"perform_target\" prefix=\"Calculate_\" suffix=\"_Tax\">
      <value>CA</value>
      <value>NY</value>
      <value>TX</value>
      ...
    </enumeration>
  </field>
</entity>
```

Then in EL:

```
perform table named (\"Calculate_\" + job.state + \"_Tax\")
```

The analyzer sees the string expression, looks up the enumeration for `job.state`, knows the bound set is `{Calculate_CA_Tax, Calculate_NY_Tax, ...}`, and records cross-table references to each. Any reference inside any of those tables is now visible. Missing entry → analyzer reports the target as **possibly unreached** (no enumeration bound) rather than treating it as silent.

### Same shape for entity-stack pushes

If we later add `entitypush <string>` or a dynamic `using` form, the bound goes on the EDD declaration the same way — declare which entity types this string is allowed to name, and the analyzer assumes any of them might be the topmost entity.

### Author burden

Authors who don't need dynamic dispatch never write enumerations — analysis stays purely static. Authors who do dynamic dispatch declare the bound once on the EDD field and the analyzer is happy. Unbound dynamic strings emit an **error** (not a warning): the analyzer can't reason about the rule set, deployment should refuse via `dtrules build --require-review`.

## Implementation pointers

- New package `pkg/dtrules/analysis/usage` to keep the regex pass and the AST pass side by side during rollout. Old pass stays exported for one release as a fallback.
- The EL parser exposes a visitor interface (`PostfixEmitter` is the existing consumer). Add a new `usage.RefCollector` that walks every DSL on every table, threading a `[]EntityType` stack as it goes.
- Cross-table reference graph: build a directed graph of \"table A's body references entity.field that came from a `for all` in caller B\" — needed because `perform`-ing into a table inherits the caller's entity stack for `executetable`-style calls.
- Wire the new analyzer into the Full Review (#768). The advisory output should distinguish:
  - **Used** — referenced somewhere reachable.
  - **Unused** — declared in EDD, no reference anywhere.
  - **Possibly used** — referenced inside an unbound-string dispatch target.
  - **Write-only** — only ever LHS of `set`, never read.

## Acceptance

- Running the new analyzer against TaxReturn produces 0 false-positive \"unused\" warnings for fields accessed via bare-name path (regression-test against TaxReturn fixtures with `for all` contexts).
- An unbound `perform table named (<expr>)` produces an error pointing at the field whose value is being computed and recommending the enumeration syntax.
- The staking project's 97 false positives collapse to whatever's genuinely unused.
- `dtrules review` surfaces the new categories.

## Closes

This will obsolete the regex pass in `analysis/edd_unused.go` once stable.

## Related

- v1.12.0 #767 (compiler advisory pass — established the advisory channel infrastructure)
- v1.12.0 #768 (Full Review — where the new analyzer's output flows)
- v1.12.0 #764 (`first pass` — example of an EL construct the static walk has to handle)
- v1.12.0 \`perform table named (<expr>)\` (the dynamic-string case the enumeration bounds address)

**Comment by PaulSnow** [comment]
Phase 1 landed in #777 (merged).

**What's in:** entity-stack-aware bare-name resolution for `for all <field>` contexts. `extractIteratedFieldReads` counts the iterating array field itself as a read. Innermost-first lookup against the EDD; conservative filtering via an EL-keyword stoplist + string-literal stripping.

**Effect on TaxReturn:** 1220 → 1193 warnings (27 former false positives now correctly counted). Staking should see a much larger collapse since its DSL is more bare-name-heavy.

**Remaining phases (this issue stays open):**

- Phase 2: `using X`, `create T as alias`, `entitypush` postfix idioms
- Phase 3: cross-table reference graph (`perform <Table>` descent)
- Phase 4: enumeration-bounded dynamic strings for `perform table named (<expr>)` — unbound dynamic perform should become a deployment-gate error
- Phase 5: replace the regex pass with a full EL AST walk

Phase 4 is the highest-value next step — it closes the static-analysis-becomes-impossible-in-the-presence-of-dynamic-dispatch loophole that the original issue called out.

**Comment by PaulSnow** [comment]
Scoping the remaining work after phases 1 and 2.

Today's state (post v1.13.0): phases 1 and 2 of this issue shipped via #777 (entity-stack-aware analysis for `for all <field>` contexts) and #778 (expanded to every push point — `using`, `for first of/in`, `for each`, `for all <type> entities`, etc.). The new analyzer is live in `pkg/dtrules/analysis/edd_unused.go`; the regex pass remains as the fall-through floor.

A narrative companion just landed in #839 (`docs/edd-usage-analyzer.md`) capturing what phases 1–2 do and what's still missing.

## What's left

Phase 3 as scoped in the issue body breaks into three independent pieces:

**Piece A — Cross-table reference graph.** A `perform <Table>` call inherits the caller's entity-stack push at the call site. Today the analyzer is per-file. Pieces: (1) build a directed graph of table→table calls during the same walk that collects references; (2) at each call edge, propagate the caller's stack state into the callee's reference resolution; (3) handle cycles (`A → B → A`) by treating the stack as a least-fixed-point over the graph.

Smallish PR — maybe 400–600 lines. Tractable in one focused PR.

**Piece B — Enumeration-bounded dynamic dispatch.** The EDD declaration syntax sketched in the issue (`<enumeration kind=\"perform_target\" prefix=\"...\" suffix=\"...\">`) needs:
- EDD parser changes to read the enumeration element
- A new `Possibly used` category in `EDDWarning`
- Compiler advisory to flag unbound `perform table named (<expr>)` as an error
- `dtrules build --require-review` to refuse on unbound dispatch

Medium PR — touches EDD loading, analyzer, advisory pass, and CLI flag. Probably 800–1000 lines, plus the EDD format spec doc update.

**Piece C — `Write-only` category split.** The current analyzer reports only `unused` and `write-only` (and they're informally distinguished by the `Reason` string). Phase 3's four-way split (`Used` / `Unused` / `Possibly used` / `Write-only`) needs an `EDDWarning` type widening and a corresponding `dtrules review` output reshape.

Smallish — couple hundred lines plus output-format test updates.

## Sequencing

The cleanest order is **A → C → B**:

- A first because cross-table propagation surfaces `unused` fields that the per-file pass missed; that needs to settle before the categorization changes.
- C next because the `Possibly used` slot needs to exist before B starts emitting things that land in it.
- B last because it's the largest piece and benefits from the cleaner output shape from C.

## Unattended decision

Pieces A and C are tractable in the unattended sweep — small enough scope, no design ambiguity. Piece B has real design decisions (the EDD enumeration syntax sketched in the issue is one option of several; the strict-error vs. warn-only behavior for unbound dispatch is a policy call).

Proceeding with **A** and **C** as separate PRs; flagging **B** here for explicit input.

For **B** the specific decisions I'd want confirmed before touching anything:

1. Is the proposed EDD syntax (`<enumeration kind="perform_target" prefix=... suffix=...>`) the one to ship, or should it be flatter (`<enumeration>` per value, with prefix/suffix computed from naming conventions)?
2. Should unbound dispatch be an **error** by default (proposed) or a **warning** with an opt-in promotion to error via `--require-review`?
3. Does the enumeration block live on the field or on the entity?

**Comment by PaulSnow** [comment]
Status after the unattended sweep.

Four of five planned pieces have landed:

| Piece | PR | What it ships |
|---|---|---|
| Piece A foundation | #841 | Static table call graph (`AnalyzeTableCallGraph`) — extracts `perform <Name>` edges, surfaces orphan calls, exposes `Callers` / `Reachable` / `UnreachedTables` helpers. Caught 4 real orphan calls in TaxReturn. |
| Piece A review wiring | #842 | Orphan calls now surface as **hard errors** in `runFullReview` / the `project_full_review` MCP tool — runtime "table not found" surprises caught at review time. |
| Piece A propagation | #844 | Cross-table entity-stack propagation via fixed-point iteration over the call graph. Bare identifiers in callee tables now resolve against the union of every caller's effective stack. Purely additive; cannot introduce false negatives. 3-level chains and cycles tested. |
| Piece C | #843 | `EDDWarning.Category` field — `unused` / `write_only` / `possibly_used` constants. Reason field stays populated for back-compat. `possibly_used` is reserved for piece B. |

## Still open

**Piece B — enumeration-bounded dynamic dispatch.** Specifically the EDD syntax sketched in the original issue body:

```xml
<field name="state" type="string">
  <enumeration kind="perform_target" prefix="Calculate_" suffix="_Tax">
    <value>CA</value>
    <value>NY</value>
    ...
  </enumeration>
</field>
```

This needs explicit input on three design choices before I touch anything:

1. **Syntax shape**: keep the proposed `<enumeration kind=... prefix=... suffix=...>` form, or go flatter (just `<enumeration><value>full_target_name</value>...` with no prefix/suffix machinery)?
2. **Error vs warning**: should an unbound `perform table named (<expr>)` be a hard error in `dtrules build --require-review` (proposed), or a warning that's only promoted to error via an explicit flag?
3. **Location**: enumeration declared on the field, or on the entity (one shared enumeration referenced by multiple fields)?

The other follow-up I noted in the original scoping comment — exposing cross-table reach in `dtrules review` output — turned out to land partially via the propagation PR (the warnings are fewer; the propagation itself is silent). A future improvement is to add a `reach_chain` field on each warning recording the caller chain that proves the field is read. That's nice-to-have, not gating; happy to pick it up if you want it but skipping it for now.

## Net effect on the staking analyzer false-positive class

Pre-piece-A: any field referenced only inside a perform-called helper looked unreferenced. Per the original issue, staking saw 97 such false positives. After #844 the propagation closes that class entirely — the only remaining "unused" warnings are either:

- Genuinely unused (declaration in EDD has no reachable reference)
- Inside a dynamic-dispatch target the analyzer can't enumerate (piece B's domain)

I don't have access to the staking project from here to measure the drop directly, but the patch shape closes the class the issue describes.

**Comment by PaulSnow** [comment]
`perform table named (<x>) with default <y>` is in, as #1137.

```
perform table named ("Determine_" + apportionment.state_code + "_Filing_Requirement")
    with default Determine_No_Filing_Requirement
```

The fallback is opt-in by existence: without the clause a missing target is the error it always was, and defining the default table is what declares "the values I didn't write a table for behave like this". The default is **named by the author** rather than derived — substituting the variable part of the concatenation with `Default` would only work for names built as a concatenation with literal ends, and would make the target depend on how the string happened to be assembled.

Tested by execution rather than compilation: the named table runs, a missing one falls back, and when neither exists the error names both.

**Two things worth recording for whoever does the enumeration work next**, since both bit during this:

- The clause had to be an *optional suffix of one alternative*, not a second alternative. As two alternatives sharing the whole prefix, ANTLR looked ahead past an arbitrarily long `strexpr` and ALL(*) prediction built state until the process was killed — ~5GB on top of an authoring save. As a suffix it costs nothing measurable (326MB against a 329MB baseline).
- `with default` is one lexer token, because `default` is a word of `RBOOLEAN`.

**What is left on this issue**, unchanged: the enumeration bound for dynamic dispatch, and with it the `possibly_used` category, which is emitted nowhere today. My suggestion from the earlier discussion still stands — derive the bound from the literal segments the analyzer already sees (`^Determine_.*_Filing_Requirement$` matches exactly the 51 state tables, with no author declaration and nothing to keep in sync) and reserve an explicit list for what derivation cannot bound. The `with default` target now also belongs in that bound, since it is a reachable dispatch target.

**Comment by PaulSnow** [comment]
The enumeration-bound piece is in, as #1155 — with the design settled in discussion rather than the one this issue sketched.

The issue proposed an enumeration declared on the EDD field. That design does not survive the corpus: the bound is a property of the **call site**, not the data — `apportionment.state_code` alone drives three differently-shaped names, and a field-level prefix/suffix has nowhere to put that. The literals already in the expression are the bound:

```
perform table named ("Determine_" + apportionment.state_code + "_Filing_Requirement")
  → ^determine_.*_filing_requirement$
```

Every defined table matching becomes a call edge; `with default X` is an edge outright; nothing is declared and nothing drifts. **Measured: CorporateTax's three dispatch sites resolve to exactly 153 edges** (51 Determine_, 102 Calculate_), zero unbounded sites in the corpus.

The refusal the issue asked for exists: no literal parts → `unbounded_dispatch` through AnalyzeExternalRefs, telling the author to anchor the name. A bound matching no table reports the same way.

**Ledger for this issue overall:** bare-name resolution fixed across three classes (#1131), `with default` (#1137), and now derived bounds with the unbounded refusal (#1155). **Still open, honestly:** the `possibly_used` category remains emitted nowhere — wiring it needs per-table reachability threaded through edd_unused, which is real plumbing; and the full AST-walk replacement of the regex pass (the issue's original scope) remains unbuilt, though the regex pass now survives hand-verification on the corpus. Both are 'next session' sized, not remainders of this one.

**Comment by PaulSnow** [comment]
## Design of record: the explicit bound (supersedes the EDD enumeration sketch)

Per Paul's direction, the bound is **stated by the author at the call site** — the same principle as `with default` (#1137): no smartness required, nothing inferred, and the declaration sits where it is read.

```
perform table named ("Determine_" + apportionment.state_code + "_Filing_Requirement")
    among Determine_CA_Filing_Requirement, Determine_NY_Filing_Requirement, ...
    with default Determine_No_Filing_Requirement
```

**Semantics:**

- `among <list>` declares the complete set of legitimate targets. The list is static, so it is checkable at compile time (each name resolves against defined tables — a listed table that does not exist is an `[external]` finding like any other `perform`).
- **Runtime enforcement is the point**: a computed name outside the list is an error naming the allowed set — dispatch cannot silently reach a table the author did not sanction. This is what a derived bound can never give: derivation describes what *could* match; the list declares what *may*.
- Composition with `with default`: a name outside the list (or inside it but with no such table) runs the default. `among` alone: outside the list is a hard error.
- Both clauses are **optional suffixes of the one grammar alternative** — the #1149 lesson; a second alternative sharing the prefix is an ALL(*) blowup.

**The derived-literal bound (#1155) stays as the fallback** for unadorned dispatch: analysis-only, over-approximating, nothing enforced. When `among` is present it wins outright — exact edges, no pattern matching.

**Analyzer:** `among` lists produce exact call edges; `possibly_used` (still unwired) becomes meaningful once reachability threads through the usage pass — unchanged scope, noted as its own piece.

**Comment by PaulSnow** [comment]
`among` is in, as #1156 — the explicit design implemented end to end.

```
perform table named ("Handle_" + job.kind)
    among Handle_AA, Handle_BB
    with default Handle_Other
```

- **Runtime enforcement**: a computed name outside the list errors naming both the computed table and the allowed set; with a default, outside-the-list or listed-but-missing runs the default.
- **Analyzer**: the list is exact call edges, and it wins outright over the derived-literal bound (#1155), which stays as the analysis-only fallback for unadorned dispatch.
- Grammar: both clauses optional suffixes of the one alternative (#1149 lesson); memory probe unchanged. Membership is RName pointer equality — case-insensitive by construction.
- Tests at every layer; docs entry with compiled postfix, gated.

**What remains on this issue** — unchanged and honestly stated: `possibly_used` needs per-table reachability threaded through the usage pass, and the AST-walk replacement of the regex pass is unbuilt. The dispatch-bounding scope — the part with design questions in it — is now complete: derived bound, explicit bound, runtime enforcement, and the unbounded refusal.

**Comment by PaulSnow** [comment]
Advanced in #1186, not closed.

**The phase-1 machinery was in place; the function feeding it returned nothing.** `initialEntities` decoded with the struct tag `initialization>initialentity`, which anchors at the document element — and every mapping in the corpus nests that block one level deeper, under `<XMLtoEDD>`. So it matched nothing, in every project, since it shipped: `baseStack` was always `[]`, and bare-name resolution never ran for any table without a `for all` context of its own.

The example its own comment cites as fixed was still broken — CHIP's `job.currentdate`, read on every run, was still reported unused. Confirmed by reverting the fix and watching it come back.

Now scanned as tokens rather than unmarshalled into a path, since depth carries no meaning here and pinning it bought nothing.

| | before | after |
|---|---|---|
| all samples, `unused` | 2687 | **2581** |
| TaxReturn | 1217 | **1123** |
| SyntaxTests | 53 | 44 |

`result.il_tax_rate` and `result.pa_tax_rate` — read bare as `result.il_taxable_income * il_tax_rate` — are no longer flagged.

**The remainder is largely genuine**, which matters as much as the reduction: 1038 of TaxReturn's remaining 1123 appear in no DSL fragment anywhere, which is what a project carrying bracket constants for unimplemented states looks like. Survivors were spot-checked rather than assumed — `result.farm_se_tax` (only `taxpayer.farm_se_tax` is read), `child_unearned_income.kiddie_tax` (only `result.kiddie_tax`), `taxpayer.estimated_payments` (its bare occurrence is `for all estimated_payments`, an iteration, not a read). All correct findings.

Still open here: the enumeration-bounded dynamic-string work, and the cross-table phases. Worth noting for whoever picks it up that the downstream consumer's 97 warnings should be re-measured against this fix before any further work — a large share of them may have been this one bug.

**Comment by PaulSnow** [comment]
Reopened — closed in error. The body of #1186 read "Does not close #776", and GitHub matched the substring `close #776` as a closing keyword on merge; its parser does not understand the negation. The enumeration-bounded dynamic-string work and the cross-table phases are still open.
